##**Polars**



Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью polars

In [28]:
import polars as pl

df = pl.read_csv('train.csv')

Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д.

In [29]:
df.describe()

shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ describe   ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       ┆ A10   ┆ C        │
│ 25%        ┆ 224.0       ┆ 0.0      ┆ 2.0      ┆ … ┆ null      ┆ 7.925     ┆ null  ┆ null     │
│ 50%        ┆ 446.0       ┆ 0.0      ┆ 3.0      ┆ … ┆ null      ┆ 14.4542   ┆ null  ┆ null     │
│ 75%        ┆ 669.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ null      ┆ 31.0      ┆ null  ┆ null     │
│ max        ┆ 891.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ WE/P 5735 ┆ 512.3292  ┆ T     ┆ S        │
└────────────┴─────────────┴──────────┴──────────┴───┴───────────┴───────────┴───────┴──────────┘

Посчитайте количество пассажиров каждого класса (Pclass)

In [30]:
class_counts = df.groupby('Pclass').agg(pl.col('Pclass').count().alias('Count'))
class_counts

<ipython-input-30-53c655f0c4f7>:1: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  class_counts = df.groupby('Pclass').agg(pl.col('Pclass').count().alias('Count'))


shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ Count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 2      ┆ 184   │
│ 3      ┆ 491   │
│ 1      ┆ 216   │
└────────┴───────┘

Выведите количество выживших мужчин и женщин на корабле

In [31]:
survived_gender_counts = df.filter(df['Survived'] == 1).groupby('Sex').agg(pl.col('Sex').count().alias('Count'))
survived_gender_counts

<ipython-input-31-b42da65fb373>:1: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  survived_gender_counts = df.filter(df['Survived'] == 1).groupby('Sex').agg(pl.col('Sex').count().alias('Count'))


shape: (2, 2)
┌────────┬───────┐
│ Sex    ┆ Count │
│ ---    ┆ ---   │
│ str    ┆ u32   │
╞════════╪═══════╡
│ female ┆ 233   │
│ male   ┆ 109   │
└────────┴───────┘

Выведите часть таблицы с пассажирами, возраст которых больше 44 лет

In [32]:
older_than_44 = df.filter(df['Age'] > 44)
older_than_44

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0    ┆ null  ┆ S        │
│             ┆          ┆        ┆ (Mary D          ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Kingcome)        ┆   ┆            ┆         ┆       ┆          │
│ 34          ┆ 0        ┆ 2      ┆ Wheadon, Mr.     ┆ … ┆ C.A. 24579 ┆ 10.5    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Edward H         ┆   ┆            ┆         ┆       ┆          │
│ …           ┆ …        ┆ …      ┆ …                ┆ … ┆ …          ┆ …       ┆ …     ┆ …        │
│ 863         ┆ 1        ┆ 1      ┆ Swift, Mrs.      ┆ … ┆ 17466      ┆ 25.9292 ┆ D17   ┆ S        │
│             ┆          ┆        ┆ Frederick Joel   ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Marg…           ┆   ┆            ┆         ┆       ┆          │
│ 872         ┆ 1        ┆ 1      ┆ Beckwith, Mrs.   ┆ … ┆ 11751      ┆ 52.5542 ┆ D35   ┆ S        │
│             ┆          ┆        ┆ Richard Leonard  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (…               ┆   ┆            ┆         ┆       ┆          │
│ 874         ┆ 0        ┆ 3      ┆ Vander Cruyssen, ┆ … ┆ 345765     ┆ 9.0     ┆ null  ┆ S        │
│             ┆          ┆        ┆ Mr. Victor       ┆   ┆            ┆         ┆       ┆          │
│ 880         ┆ 1        ┆ 1      ┆ Potter, Mrs.     ┆ … ┆ 11767      ┆ 83.1583 ┆ C50   ┆ C        │
│             ┆          ┆        ┆ Thomas Jr (Lily  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Ale…             ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

##**Ускорение работы с pandas**



Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью pandas

In [33]:
import pandas as pd

df = pd.read_csv('train.csv')

Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck

In [34]:
pip install bottleneck

In [35]:
import bottleneck as bn

mean_age = bn.nanmean(df['Age'])
print("Средний возраст пассажиров:", mean_age)

std_age = bn.nanstd(df['Age'])
print("Стандартное отклонение возраста пассажиров:", std_age)

Средний возраст пассажиров: 29.69911764705882
Стандартное отклонение возраста пассажиров: 14.516321150817317


Для каждого пассажира умножьте значение столбца Fare на 1.3, если класс его билета - 1 или 2, и на 1.1, если класс его билета - 3. Сохраните результаты как новый столбец Fare_new. (При реализации расчета используйте любые альтернативы методу iterrows)

In [36]:
df['Fare_new'] = df.apply(lambda row: row['Fare'] * 1.3 if row['Pclass'] in [1, 2] else row['Fare'] * 1.1, axis=1)
df[['Pclass', 'Fare', 'Fare_new']]

Pclass     Fare  Fare_new
0         3   7.2500   7.97500
1         1  71.2833  92.66829
2         3   7.9250   8.71750
3         1  53.1000  69.03000
4         3   8.0500   8.85500
..      ...      ...       ...
886       2  13.0000  16.90000
887       1  30.0000  39.00000
888       3  23.4500  25.79500
889       1  30.0000  39.00000
890       3   7.7500   8.52500

[891 rows x 3 columns]

##**Оптимизация типов pandas**

Считайте датасет из файла Housing.csv (это данные о ценах домов) с помощью pandas

In [37]:
df = pd.read_csv('Housing.csv')

Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [39]:
df

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

In [40]:
df.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

*Данные в колонках bedrooms, bathrooms, stories и parking лучше сделать типом int8, area типом int16, а price типом int32, тип int64 для них избыточен*

Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации

In [41]:
import numpy as np
df['price'] = df['price'].astype(np.int32)
df['area'] = df['area'].astype(np.int16)
df['bedrooms'] = df['bedrooms'].astype(np.int8)
df['bathrooms'] = df['bathrooms'].astype(np.int8)
df['stories'] = df['stories'].astype(np.int8)
df['parking'] = df['parking'].astype(np.int8)

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int32 
 1   area              545 non-null    int16 
 2   bedrooms          545 non-null    int8  
 3   bathrooms         545 non-null    int8  
 4   stories           545 non-null    int8  
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int8  
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int16(1), int32(1), int8(4), object(7)
memory usage: 35.3+ KB


*До изменения типа данных датасет весил 55.5 Kб, теперь весит 35,3 Кб*